dynamiqs/integrators/mcsolve/diffrax_integrator.py

```diff
@@ -0,0 +1,258 @@
+    jump_times: Array
+
+
+class State(eqx.Module):
+    save_state: PyTree[SaveState]
+    final_state: Array
+    final_time: float
+    num_jumps: int
+    key: Array
+
+
+def _inner_buffers(save_state):
+    assert type(save_state) is SaveState
+    return save_state.ts, save_state.ys, save_state.jump_times
+
+
+def _outer_buffers(state):
+    assert type(state) is State
+    save_state = state.save_state
+    return save_state.ts, save_state.ys, save_state.jump_times
+
+
+class MCSolveDiffraxIntegrator(MCDiffraxIntegrator, MCSolveIntegrator, SolveSaveMixin):
+    """Integrator computing the time evolution of the Monte-Carlo unraveling of the
+    Lindblad master equation using the Diffrax library."""
+
+    def _run(self, y0, tsave, rand):
+        # === prepare saveat
+        subsaveat_a = dx.SubSaveAt(ts=tsave)  # save solution regularly
+        subsaveat_b = dx.SubSaveAt(t1=True)  # save last state
+        saveat = dx.SaveAt(subs=[subsaveat_a, subsaveat_b])
+
+        # === prepare event
+        def cond_fn(t, y, *args, **kwargs):
+            return norm(y) ** 2 - rand
+
+        event = dx.Event(cond_fn, self.root_finder)
+
+        # === solve differential equation with diffrax
+        return self.diffeqsolve(
+            t0=tsave[0], t1=tsave[-1], y0=y0, saveat=saveat, event=event
+        )
+
+    def run(self):
+        # === run no jump, extract no-jump probability
+        no_jump_solution = self._run(self.y0, self.ts, 0.0)
+        no_jump_saved = jax.vmap(self.save)(unit(no_jump_solution.ys[0]))
+        final_no_jump_state = no_jump_solution.ys[1][0]
+        no_jump_prob = norm(final_no_jump_state) ** 2
+
+        # === run jump trajectories
+        jump_state_fun = jax.vmap(self._loop_over_jumps, in_axes=(0, None))
+        jump_state = jump_state_fun(self.keys, no_jump_prob)
+        jump_times = jump_state.save_state.jump_times
+        num_jumps = jump_state.num_jumps
+
+        # === save and postprocess results
+        no_jump_saved = self.postprocess_saved(
+            no_jump_saved, unit(no_jump_solution.ys[1])
+        )
+        no_jump_result = self.traj_result(
+            no_jump_saved, infos=self.infos(no_jump_solution.stats)
+        )
+        jump_saved = self.postprocess_saved(
+            jump_state.save_state.ys, jump_state.final_state[None]
+        )
+        # TODO save stats for jumps?
+        jump_result = self.traj_result(jump_saved)
+
+        return self.result(
+            no_jump_result, jump_result, no_jump_prob, jump_times, num_jumps
+        )
+
+    def _loop_over_jumps(self, key: Array, no_jump_prob: Array) -> State:
+        """loop over jumps until the simulation reaches the final time"""
+        rand_key, sample_key = jax.random.split(key)
+        rand = jax.random.uniform(rand_key, minval=no_jump_prob)
+        first_result = self._run(self.y0, self.ts, rand)
+        first_jump_time = first_result.ts[1][0]
+        time_diffs = self.ts - first_jump_time
```

Comment:
I have it this way because the loop is reserved for: applying a jump, and then solving until the next termination of `diffeqsolve` (whether that be because of an event or because `t1` is reached). I could alternatively have the loop be: solve until termination, then apply a jump. In this case we would still have a part of the evolution outside the loop, that is the final stretch from last jump to the final time. One needs to choose one or the other, and I chose the former.